libcxx/include/__iterator/reverse_iterator.h

```diff
@@ -136,10 +136,16 @@ class _LIBCPP_TEMPLATE_VIS reverse_iterator
   _LIBCPP_HIDE_FROM_ABI constexpr pointer operator->() const
     requires is_pointer_v<_Iter> || requires(const _Iter __i) { __i.operator->(); }
   {
+    // std::prev does not work for bidirectional_iterator && !LegacyBidirectionalIterator
```

Comment:
I propose that we change `flat_map::iterator`'s `iterator_category` to `random_access_iterator_tag`, so we don't have to make any changes to `reverse_iterator` or `std::prev`. This is what we did for `vector<bool>::iterator`. In all our cpp17 algorithms, we never check if they meeting the "CPP17XXXX" requirements. Instead, we simply check the `iterator_category`. We never make assumptions on whether or not the `reference` type of an iterator is actually a reference. What do you think?

Reply:
Yes, let's do that. In that case, the changes in this file can be removed from this patch.

However, we still have a general bug on `reverse_iterator` since it will mis-behave for types that are `concept bidirectional_iterator` but do not satisfy the `Cpp17BidirectionalIterator` named requirement. Specifically, based on http://eel.is/c++draft/reverse.iterators#reverse.iter.requirements, both should be acceptable for `reverse_iterator`. I think that's a separate bug on its own, and I think resolving that will mean that we don't use `std::prev` anymore from `reverse_iterator`, as your patch here does.

And then there is #109456, which is related but different since we don't even have to use `std::prev` inside `bidirectional_iterator`.

TLDR: let's create another issue to handle fixing `reverse_iterator`.

Reply:
I created a PR 
https://github.com/llvm/llvm-project/pull/112100